src/mintpy/los2slope.py

```diff
@@ -0,0 +1,373 @@
+                slope         - 1/2D np.ndarray, slope of the terrain in degree.
+                aspect        - 1/2D np.ndarray, aspect of the terrain in degree, measured in anti-clockwise direction from east.
+    Returns:    G             - 2D matrix
+    """
+    # 2D approach: not currently working, needs math re-check
+    # aspect = -aspect
+    # vh = np.sin(np.deg2rad(los_inc_angle))*np.cos(np.deg2rad(los_az_angle-aspect))
+    # vu = np.cos(np.deg2rad(los_inc_angle))
+    # ch = np.cos(np.deg2rad(slope))
+    # cu = np.sin(np.deg2rad(slope))
+    # G = vh*ch + vu*cu
+
+    # # 3d Approach
+    aspect = (
+        -aspect + 90
+    )  # Shift hillsope aspect to 0 east anti-clockwise, same as the azimuth angle.
+
+    ve = np.sin(np.deg2rad(los_inc_angle)) * np.sin(np.deg2rad(los_az_angle)) * -1
+    vn = np.sin(np.deg2rad(los_inc_angle)) * np.cos(np.deg2rad(los_az_angle))
+    vu = np.cos(np.deg2rad(los_inc_angle))
+
+    ce = (np.sin(np.deg2rad(aspect)) * np.sin(np.deg2rad(slope))) * -1
+    cn = np.sin(np.deg2rad(aspect)) * np.cos(np.deg2rad(slope))
+    cu = np.cos(np.deg2rad(slope))
+
+    G = ve * ce + vu * cu + vn * cn
+    return G
+
+
+def los2slope(dlos, los_inc_angle, los_az_angle, slope, aspect, inps, step=20):
+    """Project asc / desc LOS data onto downslope direction.
+    Parameters: dlos          - 2D np.ndarray in size of (length, width), LOS displacement in meters.
+                los_inc_angle - 1/2D np.ndarray in size of (length, width), LOS incidence angle in degree.
+                los_az_angle  - 1/2D np.ndarray in size of ( length, width), LOS azimuth   angle in degree, measured in anti-clockwise direction from east.
+                slope         - 1/2D np.ndarray, slope of the terrain in degree.
+                aspect        - 1/2D np.ndarray, aspect of the terrain in degree, measured in anti-clockwise direction from east.
+                slp_thresh    - float [inps], slope angle threshold below which the displacement is masked out
+                asp_thresh    - float [inps], hillslope aspect threshold at which aspect the displacement is masked out, usually 0.0, meaning flat location.
+                scaling_factor_thresh - float [inps], threshold for multiplication scaling factor in unit vector, this avoids very large, unrealistic values in asymptotic regions.
+                step          - int, geometry step size
+    Returns:    dslope        - 2D np.ndarray in size of (length, width), slope direction displacement in meters.
+    """
+    # initiate output
+    (length, width) = dlos.shape
+    dlos = -dlos  # convert los to make downslope positive
+    dslope = np.zeros((length, width), dtype=np.float32) * np.nan
+    scale_factor = np.zeros((length, width), dtype=np.float32) * np.nan
+
+    # 2D incidence / azimuth angle --> invert [window-by-window to speed up]
```

Comment:
do not see the need to use for loop, and do projection window-by-window as the final output can be done with matrix multiplication element-wise... this is more or the decomposition approach